pravega/pravega-operator
# Issue 312 making ss cache volumes optional

Signed-off-by: prabhaker24 Prabhaker.Saxena@dell.com

## Change log description
From Pravega 0.7, the Segment Store cache is completely in-memory. This means that the PVC that the operator creates is not necessary anymore. We need to make the deployment of this volume optional, depending on whether the version of Pravega is older or newer than 0.7, thus these changes will deploy as well as upgrade the version accordingly so as to not include pvc's incase we are deploying or upgrading to a Pravega version 0.7 or above and remove the old ones deployed by a version below 0.7 in case of an upgrade.
In Case of a rollback when an upgrade from a version above 0.7 to a version below 0.7 fails, it should remove the pods of the segment store above version 0.7 and should deploy all segment store pods of version below 0.7 along with their pvc's

## Purpose of the change
Fixes #312 

## What the code does
These changes will deploy the segment store with the pvc's for a version below 0.7 and will deploy the segment store without pvc's for a version 0.7 or above, it will also take care of upgrades in case the upgrades are from a version below 0.7 to a version above or equal to 0.7 and in this case it will remove the pvc's created by the older version while deploying the segment store for the version above 0.7.  
It will also handle the rollback in case of an upgrade failure from a version below 0.7 to version above 0.7 and will display appropriate upgrade failed message and when the user changes the version to the last version(below 0.7 from which he was upgrading) it will rollback it to this below 0.7 version and remove the old pods of the segment store having version above 0.7.
It will also display appropriate user message incase the above rollback also fails.

## How to verify it
1. Deploy Pravega with a version 0.7 or above and it should be deployed without any pvc's
2. Upgrade Pravega with a version blow 0.7 to a version above 0.7 it should do two this deploy the new segment store without pvc's and remove the old segment store as well as the pvc's attached to it.
3. Upgrade Pravega from a version below 0.7 to a version above 0.7 
4. Upgrade Pravega from a version above 0.7 to a version above 0.7
5. Upgrade from 0.6.0 to 0.6.2 . 
6. I/o should be working in case of all the above four processes both while the upgrade or deployment is in the process as well as when it's completed to check that Benchmark can be run.
7. Rollback in case the upgrade fails while upgrading from a version below 0.7 to a  version below 0.7
8.  Rollback in case the upgrade fails while upgrading from a version below 0.7 to a  version above 0.7
9. Rollback in case the upgrade fails while upgrading from a version above 0.7 to a  version above 0.7
10. Displays appropriate message in case rollback fails while doing a rollback if an upgrade fails  from a version below 0.7 to a  version above 0.7 .
